nemo/collections/nlp/models/language_modeling/megatron_base_model.py

```diff
@@ -322,9 +323,33 @@ def _reconfigure_val_batches(self):
         """
         Reconfigure trainer.limit_val_batches for pretraining
         """
+        # Override limit_val_batches to be a multiple of num microbatches and so there are limit_val_batches//num_micro_batches num of global batches
         if isinstance(self.trainer.limit_val_batches, int):
-            # Override limit_val_batches to be a multiple of num microbatches and so there are limit_val_batches//num_micro_batches num of global batches
             self.trainer.limit_val_batches *= get_num_microbatches()
+        else:
+            assert isinstance(self.trainer.limit_val_batches, float)
+            # Don't reconfigure if limit_val_batches is 0.0
+            if self.trainer.limit_val_batches == 0.0:
+                return
+            if self._validation_ds is not None and len(self._validation_dl) != float("inf"):
```

Comment:
How do infinite validation sets work in nemo? Is this even a valid approach? Shouldn't this be caught on dataset on argument layer, even before dataset construction?

Reply:
I ported that condition check over from PTL as they check if length of val dataloader is not inf [here](https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/trainer/connectors/data_connector.py#L449) while casting float limit_val_batches to int.
I don't think NeMo has any cases for infinite validation sets.

Reply:
The code now does the round down logic only if limit_val_batches > 0.0 and less than 1.0.

Reply:
As I understand this, we have 4 cases if `limit_val_batches` is a float:

1.  `limit_val_batches` == 0.0. Then we stick with this value and let PTL skip validation
2. `limit_val_batches` == 1.0. Then we set `self.trainer.limit_val_batches = len(self._validation_dl)`, even it's not divisible by `get_num_microbatches()`
3. `limist_val_batches * len(self._validation_dl) < 1`. Then we raise an exception
4. `limist_val_batches * len(self._validation_dl) > 1`. Then we round down to the integer multiple of `get_num_microbatches()`

@athitten do I understand this correctly? Doesn't it hang if `limit_val_batches == 1.0`?

Reply:
Nope, it does not hang when `limit_val_batches == 1.0`. When l`imit_val_batches == 1.0`, we use the full len of the dataloader as the `limit_val_batches` and since the `len(dataloader)` is already in microbatches, it is divisible by `get_num_microbatches()`. The situation where it needs to be ensured to be a multiple of  `get_num_microbatches()`, arises when `0<limit_val_batches<1.0`. Basically a fraction.